### PR TITLE
[SYCL-MLIR]: Enable reduction detection at -O1 and above

### PR DESCRIPTION
**Pull request by etiotto**

### CODE REVIEW

File: polygeist/tools/cgeist/driver.cc [hunk 1]
```diff
@@ -356,10 +356,10 @@ static int optimize(mlir::MLIRContext &Ctx,
   GreedyRewriteConfig CanonicalizerConfig;
   CanonicalizerConfig.maxIterations = CanonicalizeIterations;
 
-  if (DetectReduction)
-    OptPM.addPass(polygeist::detectReductionPass());
-
   if (OptimizationLevel != llvm::OptimizationLevel::O0) {
+    if (DetectReduction)
+      OptPM.addPass(polygeist::detectReductionPass());
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/driver.cc:361]: Changed the PR title, as it is enabled at O1 and above.

File: polygeist/tools/cgeist/Test/Verification/redstore.c [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// RUN: cgeist %s --function=* --detect-reduction -S | FileCheck %s
+// RUN: cgeist -O2 %s --function=* -S | FileCheck %s
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/redstore.c:1]: could use `-O1` (same for other modified test cases)

**Review by victor-eds — APPROVED** [review]

File: polygeist/tools/cgeist/Test/Verification/redstore.c (hunk 2 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/redstore.c:1]: Well I think -O2 is what the `clang` driver pass down to cgeist by default when we compiler SYCL code so ...

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/redstore.c:1]: Sure.

**Review by whitneywhtsang — CHANGES_REQUESTED** [review]
There are some test suite failures need to be fixed:
L0 GEN9:
```
  Failed Tests (6):
    SYCL :: Plugin/interop-level-zero-buffer-ownership.cpp
    SYCL :: Plugin/interop-level-zero-buffer.cpp
    SYCL :: Plugin/interop-level-zero-get-native-mem.cpp
    SYCL :: USM/allocator_container.cpp
    SYCL :: USM/allocator_vector.cpp
    SYCL :: USM/smem_concurrent.cpp
```
OCL GEN9:
```
  Failed Tests (2):
    SYCL :: USM/allocator_container.cpp
    SYCL :: USM/allocator_vector.cpp
```
OCL x64:
```
  Failed Tests (2):
    SYCL :: USM/allocator_container.cpp
    SYCL :: USM/allocator_vector.cpp
```

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**